src/pybennu/pybennu/providers/power/solvers/siren.py

```diff
@@ -0,0 +1,493 @@
+logging.addLevelName(LEVELS['filter'], 'FILTER')
+logging.basicConfig(format='%(asctime)s %(name)-12s  %(levelname)-8s  %(message)s', level=logging.DEBUG, datefmt='%Y-%m-%d %H:%M:%S')
+
+class CircularBuffer:
+    def __init__(self, max_size):
+        self.max = max_size
+        self.cur_size = 0
+        self.buffer = deque()
+        self.sum = 0
+
+    def add(self, value):
+        if len(self.buffer) == self.max:
+            self.sum -= self.buffer.popleft()
+
+        self.buffer.append(value)
+        self.sum += value
+        self.cur_size = len(self.buffer)
+            
+            
+
+
+
+class device_config:
+    def __init__(self, _name):
+        self.name = _name
+        self.polling_time = 0.5
+        self.type = ""
+        self.id = ""
+        self.config_file = ""
+        self.server_endpoint = "tcp://127.0.0.1:5555"
+        self.publish_endpoint = "udp://239.0.0.1:40000"
+        self.filter = "/"
+        self.level_name = ''
+        self.to_endpoint = {}
+        self.to_device = {}
+
+
+class Siren(Provider):
+    def __init__(self, server_endpoint, publish_endpoint, settings):
+        """
+        Args:
+            server_endpoint: IP address for the provider server to listen on
+            publish_endpoint: IP address for the provider to publish values to
+            settings: (PybennuSettings) The settings found in the siren_config.yaml 
+                      file generated from the scenario yaml
+        """
+        super().__init__(server_endpoint, publish_endpoint)
+
+        self.conf: PybennuSettings = settings
+        self.setup_logger()
+
+        siren_config = self.conf.siren.siren_json
+        passed_devices = None
+        self.setup_siren_config(siren_config, passed_devices) # Populates self.device_configs (dict)
+
+        self.devices = {}
+        self.to_endpoint_states = {} # Holds hardware values before written to internal_state
+        self.to_hardware_threads = []
+        self.from_hardware_threads = []
+
+        for device_name, dev_config in self.device_configs.items():
+
+            self.devices[device_name] =  self.create_device(dev_config)
+
+            self.to_endpoint_states[device_name] = dict.fromkeys(dev_config.to_endpoint.keys())
+
+            # Writes values to hardware from internal state.
+            # Replaces _subscription_handler functionality
+            if dev_config.to_device.keys():
+                thread = threading.Thread(target=self.write_to_hardware, args=(device_name,)) 
+                thread.daemon = True
+                self.to_hardware_threads.append(thread)
+
+            # Writes value to internal state from hardware
+            # Replaces _to_provider functionality
+            if dev_config.to_endpoint.keys():
+                thread = threading.Thread(target=self.update_from_hardware, args=(device_name,))
+                thread.daemon = True
+                self.from_hardware_threads.append(thread)
+
+        self.run_thread = True
+
```

Comment:
remove extraneous newlines, there should be no more than one blank line separating most code. Follow PEP8. one of these days we'll add a linter to make this easier...